File: datastruct/algrithm/dp/all_parentheses.cpp
#include <iostream>
using namespace std;
void all_p(int i,int j)
{
    int k=0;
    if (i==j)
        std::cout << "A"<<i; 
    else{
        std::cout << "(";
        for (k = i; k < j; ++k) {
            all_p(i,k);
            all_p(k+1,j);
        }
        cout<< ")";
    }

}
int main(void)
{
    all_p(1,4);
    return 0;
}

File: datastruct/algrithm/dp/all_parentheses.py
def all_p(i,j):
    if i==j:
        print "A"+str(i),
    else:
        #import pdb; pdb.set_trace()
        for k in range(i,j):
            print "(",
            all_p(i,k)
            all_p(k+1,j)
            print ")",
        print "+",

stra=" "
#def all_p(i,j,str):
    #if i==j:
        #stra=stra+ "A"+str(i),
    #else:
        #for k in range(i,j):
            #stra=stra+"("
            #all_p(i,k)
            #all_p(k+1,j)
            #stra=stra+")"

if __name__ == "__main__":
    #all_p(1,4,stra)
    all_p(1,4)
